fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Utils.java

```diff
@@ -536,4 +569,273 @@ private static <T extends ScalarOperator> void collect(ScalarOperator root, Clas
 
         root.getChildren().forEach(child -> collect(child, clazz, output));
     }
+
+    public static void getRelatedMvs(List<Table> tablesToCheck, Set<MaterializedView> mvs) {
+        Set<MvId> newMvIds = Sets.newHashSet();
+        for (Table table : tablesToCheck) {
+            Set<MvId> mvIds = table.getRelatedMaterializedViews();
+            if (mvIds != null && !mvIds.isEmpty()) {
+                newMvIds.addAll(mvIds);
+            }
+        }
+        if (newMvIds.isEmpty()) {
+            return;
+        }
+        List<Table> newMvs = Lists.newArrayList();
+        for (MvId mvId : newMvIds) {
+            Database db = GlobalStateMgr.getCurrentState().getDb(mvId.getDbId());
+            if (db == null) {
+                continue;
+            }
+            Table table = db.getTable(mvId.getId());
+            if (table == null) {
+                continue;
+            }
+            newMvs.add(table);
+            mvs.add((MaterializedView) table);
+        }
+        getRelatedMvs(newMvs, mvs);
+    }
+
+    // get all ref tables within and below root
+    public static List<Table> getAllTables(OptExpression root) {
+        List<Table> tables = Lists.newArrayList();
+        getAllTables(root, tables);
+        return tables;
+    }
+
+    private static void getAllTables(OptExpression root, List<Table> tables) {
+        if (root.getOp() instanceof LogicalScanOperator) {
+            LogicalScanOperator scanOperator = (LogicalScanOperator) root.getOp();
+            tables.add(scanOperator.getTable());
+        } else {
+            for (OptExpression child : root.getInputs()) {
+                getAllTables(child, tables);
+            }
+        }
+    }
+
+    public static boolean isValidMVPlan(OptExpression root) {
+        return isLogicalSPJ(root) || isLogicalSPJG(root);
+    }
+
+    public static boolean isLogicalSPJG(OptExpression root) {
```

Comment:
Please consider use the `OptExpressionVisitor` to check it, which makes it more extensible.

Later we need a similar `PhyscialSPJG` check.

Reply:
May we could use rule pattern to check this?